### ISSUE TITLE
Multi-strip usage when encoding group4 TIF breaks use in PDF

### ISSUE THREAD

**Issue by josch** [python-pillow/Pillow#5740]
Hi,

since #5514, group4 tiff images are not anymore saved as a single strip. This breaks the use-case of utilizing PIL for encoding images as group4 for inclusion in PDF documents, as the PDF format only supports a single group4 encoded strip per image.

Please allow for a way to force single-strip group4 encoding to restore this functionality, maybe by allowing to specify the maximum strip length?

(I realize this sounds a lot like https://xkcd.com/1172/ :smile:)

**Comment by kmilos** [comment]
Perhaps adding a [global constant that can be overridden instead of the hard-coded 64 KB here](https://github.com/python-pillow/Pillow/blob/9025ae8332fd0348e9a1c1552d5f9a1f3b765ba1/src/PIL/TiffImagePlugin.py#L1622) could be a neat way to address this?

**Comment by josch** [comment]
@radarhere thanks a lot for your input! Yes, that hack is indeed interesting to me because even if @kmilos solution is implemented in a future pillow version, we cannot travel back in time and thus I would need a solution for those pillow versions that have #5514 applied but do not yet have a constant that can be overridden.

Unfortunately, I'm afraid that I'm not getting your code snippet to work. Did you test it? I suspect that how it is written, the values will not be overwritten by the IFD writer.

**Comment by kmilos** [comment]
> Unfortunately, I'm afraid that I'm not getting your code snippet to work.

Perhaps `STRIPBYTECOUNTS` and `STRIPOFFSETS` need to be tuples...

**Comment by josch** [comment]
> > Unfortunately, I'm afraid that I'm not getting your code snippet to work.
> 
> Perhaps `STRIPBYTECOUNTS` and `STRIPOFFSETS` need to be 1-tuples, e.g. (31500,) and (0,) in the example above...

Yes, I tried that already and the resulting image is corrupted.

**Comment by radarhere** [comment]
I've tested the code I wrote, yes.

Could we get a copy of your original image, and the values you're trying to write?

**Comment by josch** [comment]
> I've tested the code I wrote, yes.
> 
> Could we get a copy of your original image, and the values you're trying to write?

Sorry, it was my mistake. Of course I have to reset `TiffImagePlugin.ImageFileDirectory_v2.__getitem__` back to its original before reading the bytecounts and offsets so that I can extract the CCITT group4 data. It works now.

Thanks a lot! :)

**Comment by radarhere** [comment]
To explicitly point it out here, #5744 since added the ability to set `STRIP_SIZE`.

You might also be interested to know that #6453 has requested group4 compression be a part of Pillow itself. I've created PR #6470 to add this feature.

**Comment by radarhere** [comment]
#6470 has been merged, so `strip_size` can now also be passed in as an encoder argument when saving - `im.save("out.tif", strip_size=...)`

**Comment by josch** [comment]
Thank you @radarhere, using `im.save("out.tif", strip_size=...)` will be much cleaner than temporarily setting `PIL.TiffImagePlugin.STRIP_SIZE` which has the potential of leaving the program in a bad state if the execution flow happens to leave before the original value is restored (for example due to an exception).

But is there a way (other than checking and comparing the PIL version) to check whether the current version of PIL supports `im.save("out.tif", strip_size=...)`? With `PIL.TiffImagePlugin.STRIP_SIZE` I was able to check `hasattr(TiffImagePlugin, "STRIP_SIZE")` and only then execute the code that sets `PIL.TiffImagePlugin.STRIP_SIZE`. Can a check like that also be done for whether an image plugin supports a specific encoder argument?

**Comment by radarhere** [comment]
Not directly, no.

But if you don't like version numbers for some reason, perhaps you could check indirectly.

```python
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image
>>> Image.init()
1
>>> "MPO" in Image.SAVE_ALL
True
```
#6444 added support for saving multiple MPO frames, as will be released in Pillow 9.3.0 for the first time. So by checking if Pillow supports saving multiple MPO frames, you are checking that Pillow is >= 9.3.0 (presuming that you aren't concerned about people compiling Pillow from source in between versions).

**Comment by milahu** [comment]
> extract the CCITT group4 data

example code fragment
used in https://github.com/milahu/binarize-pdf/commit/5a9ca491771b1cbd8e83c85b5d0201763e5312ab

```py
import numpy as np
import io
from PIL import Image
import tifffile
import math

# np_img = np.array(pil_img)
np_img = get_image_as_np_array()

# CCITT Group 4 compression
pil_img = Image.fromarray(np_img, mode="L").convert("1")
tiff_buf = io.BytesIO()
# produce a single strip
strip_size = math.ceil(pil_img.width / 8) * pil_img.height
pil_img.save(
    tiff_buf,
    format="TIFF",
    compression="group4",
    strip_size=strip_size,
)
tiff_buf.seek(0)
with tifffile.TiffFile(tiff_buf) as tiff:
    page = tiff.pages[0]
    assert len(page.dataoffsets) == 1 # require single strip
    offset = page.dataoffsets[0]
    size = page.databytecounts[0]
tiff_buf.seek(offset)
ccitt_bytes = tiff_buf.read(size)
del tiff_buf

# PDF output
filters.append("/CCITTFaxDecode")
# height, width = np_img.shape[:2]
height, width = pil_img.height, pil_img.width
DecodeParms = (
    b"<<\n"
    b"      /K -1\n" +
    f"      /Columns {width}\n".encode("ascii") +
    f"      /Rows {height}\n".encode("ascii") +
    b"      /BlackIs1 true\n"
    b"   >>"
)
```
